fix(expo-location): Report location updates immediately in foreground on Android

tyrauber opened the pull request:
# Why

On Android, deferredUpdates were being applied to foreground operations due to incorrect mIsHostPaused initialization. Unlike iOS, where distanceInterval and timeInterval are respected in foreground, Android was batching all location
updates using the deferred updates logic regardless of app state.

This caused significant delays before location updates appeared on screen, resulting in poor UX for location tracking apps.

# How

Root Cause:

The mIsHostPaused flag was initialized to true. Since apps start in foreground (lifecycle callbacks fire after initialization), all location updates were routed through the deferred updates path from the start.

Solution:
1. Initialize mIsHostPaused = false since apps start in foreground state
2. Add reportLocationsImmediately() method that bypasses the deferred buffer
3. Route location updates based on foreground/background state:
   - Foreground: report immediately (matches iOS behavior)
   - Background: use deferred buffer for battery optimization

This brings Android behavior in line with iOS, respecting distanceInterval and timeInterval in foreground while preserving deferred batching in background.

# Test Plan

- Verified immediate location delivery in foreground (matches iOS)
- Confirmed background deferred batching still works for battery optimization
- Validated foreground/background transitions work correctly

# Checklist


- [X] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [X] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [X] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

tyrauber committed:
[android][location] Fix deferred updates being applied in foreground

On Android, deferredUpdates were being applied to foreground operations due to
incorrect mIsHostPaused initialization. Unlike iOS, where distanceInterval and
timeInterval are respected in foreground, Android was batching all location
updates using the deferred updates logic regardless of app state.

This caused significant delays before location updates appeared on screen,
resulting in poor UX for location tracking apps.

Root Cause:
The mIsHostPaused flag was initialized to true. Since apps start in foreground
(lifecycle callbacks fire after initialization), all location updates were
routed through the deferred updates path from the start.

Solution:
1. Initialize mIsHostPaused = false since apps start in foreground state
2. Add reportLocationsImmediately() method that bypasses the deferred buffer
3. Route location updates based on foreground/background state:
   - Foreground: report immediately (matches iOS behavior)
   - Background: use deferred buffer for battery optimization

This brings Android behavior in line with iOS, respecting distanceInterval and
timeInterval in foreground while preserving deferred batching in background.

Testing:
- Verified immediate location delivery in foreground (matches iOS)
- Confirmed background deferred batching still works for battery optimization
- Validated foreground/background transitions work correctly

tyrauber committed:
Update packages/expo-location/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

tyrauber committed:
chore: update changelog with PR#

chrfalch reviewed (changes requested):
Hi, @tyrauber - thanks again for your contribution! I've reviewed it and have a change request to make the code even clearer. 

In addition - could you run this command so that we can fix the failing Android CI job?:

`./gradlew :expo-location:spotlessApply`

(Run it in the apps/bare-expo/android folder f.ex).

tyrauber committed:
[android][location] Refactor handleLocationUpdate to consolidate foreground/background logic

Extract foreground vs background decision into handleLocationUpdate() method
for cleaner code structure. The method now takes locations as parameter and
decides internally whether to report immediately (foreground) or defer
(background), eliminating duplicate conditional checks in didReceiveBroadcast.

tyrauber committed:
Fix: linting with (./gradlew :expo-location:spotlessApply in apps/bare-expo/android)

lukmccall reviewed (approved):
Looks good! Thanks for your contribution
